libs/platform/ebpf_serialize.c

```diff
@@ -179,3 +214,384 @@ ebpf_deserialize_map_information_array(
 
     return result;
 }
+
+void
+ebpf_program_information_free(_Pre_maybenull_ _Post_invalid_ _In_ ebpf_program_information_t* program_info)
+{
+    if (program_info != NULL) {
+        ebpf_free(program_info->program_type_descriptor.context_descriptor);
+        ebpf_free((void*)program_info->program_type_descriptor.name);
+        for (uint32_t i = 0; i < program_info->count_of_helpers; i++) {
+            ebpf_helper_function_prototype_t* helper_prototype = &program_info->helper_prototype[i];
+            ebpf_free((void*)helper_prototype->name);
+        }
+        ebpf_free(program_info->helper_prototype);
+        ebpf_free(program_info);
+    }
+}
+
+ebpf_result_t
+ebpf_serialize_program_information(
+    _In_ const ebpf_program_information_t* program_info,
+    _Out_writes_bytes_to_(output_buffer_length, *serialized_data_length) uint8_t* output_buffer,
+    size_t output_buffer_length,
+    _Out_ size_t* serialized_data_length,
+    _Out_ size_t* required_length)
```

Comment:
Why do you need both of these?  Isn't serialized_data_length always 0 on failure and required_length on success?

Reply:
The IOCTLs all use protocol request/reply messages which have a header field with length field - that shows the length of the entire message.
If the reply message is of variable length, it has an additional size field (such as in map_information and program_information reply cases). Given this, I thought it was convenient to return 2 output size parameters, specially in the failure case, as reply->size will get filled with 0, whereas required_length could be used to populate the length. Yes, in success case they both will be same.
If you see any issue with this, I can file a new github issue and address it in a separate PR.